### PR TITLE
Support large model export using multi-gpu

### PR DESCRIPTION
**Pull request by wejoncy**
### Description

This PR is to implemente a exporter which works for large language models(LLM).
It works for models like Llama2-70b or gpt-175.

The main idea is to utilize multiple-GPU and dispatch differnet layers to different GPU, in short, it symply implemented auto pipeline parallelism.

For example : to export Llama2-70b, you need 8x V100-32GB or 4x A100-80G or More GPU memories.

It would expect to export decoder-only models. For encoder-decoder arch-like models, we didn't test it yet.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+import torch
+import transformers
+
+
+class Exporter:
+    """
+    A class for exporting large transformer models to ONNX format.
+    """
+
+    def __init__(self):
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:16]: The exporter class has a single state model. Would it make sense to move the methods out as functions for readers to reason about them more easily?

Comment (wejoncy) [reply onnxruntime/python/tools/transformers/large_model_exporter.py]: Thanks. Make sense

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    buffer_size = 0
+    buffer_sum = 0
+    for buffer in model.buffers():
+        buffer_size += buffer.nelement() * buffer.element_size()
+        buffer_sum += buffer.nelement()
+    all_size = (param_size + buffer_size) / 1024 / 1024
+    return all_size
+
+
+def model_prepare(hf_model: str, tokenizer=None):
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:50]: nit: optional: naming: (fyi)

Prefer verb phrases (`prepare_model`) as function names. Better yet, prefer a more descriptive name that describes what the function does

```suggestion
def initialize_model_and_sample_inputs(hf_model: str, tokenizer=None):
```

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    for all models
+    """
+    user_inputs = []
+
+    def hook_for_inputs(_, inputs, kwargs):
+        user_inputs.append((inputs, kwargs))
+        return user_inputs[0]
+
+    hook_handle = model.register_forward_pre_hook(hook_for_inputs, with_kwargs=True)
+    import inspect
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:150]:
```suggestion
```

prefer importing at the top of the module when possible so we don't hit import errors deep in the stack.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+    sample_inputs = tuple(tokenizer("Hello, my dog is cute", return_tensors="pt").values())
+    return onnx_model_name, model, sample_inputs
+
+
+def auto_pipeline(model: nn.Module, gpulist: list, sample_inputs: tuple):
+    """
+    make a model can be executed across multile-gpu.
+    it's a pipeline method
+    """
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:75]: nit: 

```suggestion
    """Make the model executable across multiple gpus."""
```

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    all_size = (param_size + buffer_size) / 1024 / 1024
+    return all_size
+
+
+def model_prepare(hf_model: str, tokenizer=None):
+    """
+    prepare torch model, name, and inputs
+    """
+    onnx_model_name = Path(hf_model + "/").name
+    import re
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:55]:
```suggestion
```

nit: move to top

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+import onnx
+import torch
+import transformers
+from torch import nn
+
+
+def disable_huggingface_init():
+    # do not init model twice as it slow initialization
+    import torch
+    import torch.nn.init
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:23]:
```suggestion
```

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    onnx_inp_names = ("input_ids", "attention_mask")
+    onnx_out_names = ("logits",)
+    onnx_dynamic_axes = {
+        "input_ids": {0: "batch_size", 1: "seq_len"},
+        "attention_mask": {0: "batch_size", 1: "seq_len"},
+    }
+    torch.onnx.export(
+        model=model,
+        args=onnx_inputs,
+        f=str(onnx_filepath_export_multi_files_tmp),
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:225]:
```suggestion
        f=os.fspath(onnx_filepath_export_multi_files_tmp),
```

Use https://docs.python.org/3/library/os.html#os.fspath for correctly obtaining the path.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        verbose=False,
+        opset_version=16,
+        input_names=onnx_inp_names,
+        output_names=onnx_out_names,
+        dynamic_axes=onnx_dynamic_axes,
+    )
+
+    onnx_model = onnx.load(str(onnx_filepath_export_multi_files_tmp))
+
+    onnx_path.exists() and onnx_path.unlink()
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:235]: Why is this needed? Will save_model simple overwrite the file?

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    (onnx_path.parent / f"{onnx_model_name}_ext.data").exists() and (
+        onnx_path.parent / f"{onnx_model_name}_ext.data"
+    ).unlink()
```
Comment (justinchuby) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:238]: If this is needed, perhaps https://docs.python.org/3/library/pathlib.html#pathlib.Path.unlink:

```suggestion
    (onnx_path.parent / f"{onnx_model_name}_ext.data").unlink(missing_ok=True)
```

File: onnxruntime/python/tools/transformers/large_model_exporter.py (hunk 7 above)
Comment (wejoncy) [reply onnxruntime/python/tools/transformers/large_model_exporter.py:225]: onnx_filepath_export_multi_files_tmp is Path variable, so we are just convert it to string. 
Not sure "os.fpath" can work here.

The reason I have to convert to str here is that torch.onnx.export doesn't work for Path and it wouldn't report a typical error.

File: onnxruntime/python/tools/transformers/large_model_exporter.py (hunk 8 above)
Comment (wejoncy) [reply onnxruntime/python/tools/transformers/large_model_exporter.py:228]: Onnx.export has some `bugs` when there is a same-name file.

The new model will be append after the existed file.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+    # input_keys would be usesful if the model has some special inputs
+    input_keys, onnx_inputs = retrieve_onnx_inputs(model, sample_inputs)
+
+    onnx_path: Path = Path(onnx_path_str).absolute()
+    if onnx_path.suffix != ".onnx":
+        onnx_path = onnx_path / onnx_model_name
+
+    onnx_filepath_export_multi_files_tmp: Path = onnx_path.parent / "tmp/tmp.onnx"
+    onnx_filepath_export_multi_files_tmp.parent.exists() and shutil.rmtree(onnx_filepath_export_multi_files_tmp.parent)
```
Comment (thiagocrepaldi) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:210]: Maybe https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile can help managing temporary files more easily

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    onnx_inp_names = ("input_ids", "attention_mask")
+    onnx_out_names = ("logits",)
+    onnx_dynamic_axes = {
+        "input_ids": {0: "batch_size", 1: "seq_len"},
+        "attention_mask": {0: "batch_size", 1: "seq_len"},
+    }
+    torch.onnx.export(
+        model=model,
+        args=onnx_inputs,
+        f=str(onnx_filepath_export_multi_files_tmp),
```
Comment (thiagocrepaldi) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:222]: `tempfile.NamedTemporaryFile.name` is a string already

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        verbose=False,
+        opset_version=16,
```
Comment (thiagocrepaldi) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:224]: maybe 17?

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    get the pretrained torch model from hugginface,
+    and onnx_model_name, sample model-inputs
+    """
+    onnx_model_name = Path(hf_model + "/").name
+
+    onnx_model_name = re.sub(r"[^0-9a-zA-Z]", onnx_model_name, "_") + ".onnx"
+    disable_huggingface_init()
+
+    model = transformers.AutoModelForCausalLM.from_pretrained(  # type: ignore
+        hf_model, torch_dtype=torch.float16, trust_remote_code=True
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:63]: Maybe be useful to add cache_dir=... (and also `--cache_dir` in argument). It is because the model is very large, default cache directory (usually the OS disk) might not have enough space for large model.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+        return (tuple(modifyed_inputs), kwargs)
+
+    def move_layer_to_device_rurc(mod, dev):
+        mod.to(dev)
+        for layer in mod.named_children():
+            move_layer_to_device_rurc(layer[1], dev)
+
+    model = model.half()
+    all_hooks = []
+    # model.register_module_forward_pre_hook(input_gpu_device_hook)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:108]: remove the commented code.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    sample_inputs_tp: inputs for torch model
+    """
+    onnx_model_name, model, sample_inputs_tp = initialize_model_and_sample_inputs(hf_model)
+    total_mem_per_cpu = torch.cuda.get_device_properties(0).total_memory / 1024 / 1024
+
+    print("Model_Size", get_model_parameter_size(model))
+    print("total_mem_per_cpu=", total_mem_per_cpu)
+    if get_model_parameter_size(model) > total_mem_per_cpu * 0.45:
+        if torch.cuda.device_count() > 1:
+            print("multi-gpu")
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:185]: May print the number of GPUs here.

File: onnxruntime/python/tools/transformers/large_model_exporter.py [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+
+    onnx_filepath_export_multi_files_tmp: Path = onnx_path.parent / "tmp/tmp.onnx"
+    onnx_filepath_export_multi_files_tmp.parent.exists() and shutil.rmtree(onnx_filepath_export_multi_files_tmp.parent)
+    os.makedirs(onnx_filepath_export_multi_files_tmp.parent)
+
+    onnx_inp_names = ("input_ids", "attention_mask")
+    onnx_out_names = ("logits",)
+    onnx_dynamic_axes = {
+        "input_ids": {0: "batch_size", 1: "seq_len"},
+        "attention_mask": {0: "batch_size", 1: "seq_len"},
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/large_model_exporter.py:217]: Shall we add past/present key/value inputs/outputs?

File: onnxruntime/python/tools/transformers/large_model_exporter.py (hunk 12 above)
Comment (wejoncy) [reply onnxruntime/python/tools/transformers/large_model_exporter.py:224]: Sure

File: onnxruntime/python/tools/transformers/large_model_exporter.py (hunk 16 above)
Comment (wejoncy) [reply onnxruntime/python/tools/transformers/large_model_exporter.py]: Done

**Review by yufenglee — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]
